### DISCUSSION TITLE
Cost of evaluating the two branches of jnp.where

### DISCUSSION THREAD

**Discussion by fbartolic** [jax-ml/jax#10306] [Q&A] [answered]
Consider the following function

```Python
import jax.numpy as jnp

@jit 
def f(x):
    cond = g(x)
    return jnp.where(cond, A(x), B(x))
```
where  branch B is at least 1000x more expensive to evaluate than A. My understanding is that in a jit-ed function both branches are evaluated if `cond` is `True` but branch B is evaluated with using arrays with abstract values and no floating point operations are executed. If that's the case I'd expect that evaluating `f(x)` for values where `g(x) == True` is substantially cheaper than evaluating `B(x)` but I'm not getting any noticeable speedup. What's going on here?

**Comment by soraros** [comment]
`jnp.where` is lowered into `select` which evaluates both branches. You could try `lax.cond`.

**Comment by YouJiacheng** [comment, marked as answer]
`jnp.where` always concretely evaluate both branches, and the output can be a mixture of two branches if `cond` is an array.
You can use `lax.cond` for this case, which only allows scalar condition and is lazy evaluated.
`lax.cond` do exactly what you want: both branches are traced to jaxpr, while only execute one of them in runtime.

**Reply by fbartolic** [reply]
Ohhhh, I did not know this. I've been using `jnp.where` everywhere thinking that it works the way you described `lax.cond` working. I've just tried `lax.cond` and it fixed the problem.  With `lax.cond` evaluating `f` is ~5x more expensive than evaluating `A(x)` for `cond == True`. Thank you!!!

**Reply by YouJiacheng** [reply]
One more thing: `lax.cond` can be slow / have a (relative) large overhead on accelerator(GPU/TPU), you should be careful when there are a lots of branches(e.g. using `lax.cond` inside `vmap` and `scan`), or make sure the branches computation cost is large enough.

**Reply by jakevdp** [reply]
One minor detail to be aware of: the batching rule for `cond` is implemented in terms of `select`, so as soon as you `vmap` an expression with `cond`, you will always evaluate both branches https://github.com/google/jax/blob/80a73efc33907fc3493fc7f53a31da36f8ed20cb/jax/_src/lax/control_flow.py#L991

**Reply by jakevdp** [reply]
That's a good idea - I'm not sure where this would best be documented. Do you have any thoughts?

**Reply by YouJiacheng** [reply]
Oops, thanks for pointing this out, I haven't known that before! IIRC, this behavior isn't documented, would you add some comments to docs? @jakevdp

**Reply by fbartolic** [reply]
@jakevdp thanks for pointing that out, I'll have avoid using `vmap` with this function then. I would also appreciate seeing more documentation on this :).

**Reply by YouJiacheng** [reply]
Maybe add a note in `lax.cond` document is enough. and it can be better if also add some note in sharp bits about control flow and vmap.

**Reply by soraros** [reply]
@jakevdp I'd vote for somewhere in [The Sharp Bits - Control Flow](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#control-flow).
